src/proxy/http/fallback.rs

```diff
@@ -0,0 +1,336 @@
+use bytes::Buf;
+use futures::{Async, Future, Poll};
+use http;
+use hyper::body::Payload;
+use proxy;
+use svc;
+
+use std::{fmt, marker::PhantomData, mem};
+
+#[derive(Debug)]
+pub struct Error<A> {
+    error: proxy::Error,
+    fallback: Option<http::Request<A>>,
+}
+
+pub type Layer<P, F, A> = MakeFallback<svc::Builder<P>, svc::Builder<F>, A>;
+
+#[derive(Debug)]
+pub struct MakeFallback<P, F, A> {
+    primary: P,
+    fallback: F,
+    _p: PhantomData<fn(A)>,
+}
+
+pub struct Service<P, F, A> {
+    primary: P,
+    fallback: F,
+    _p: PhantomData<fn(A)>,
+}
+
+pub struct MakeFuture<P, F, A>
+where
+    P: Future,
+    F: Future,
+{
+    primary: Making<P>,
+    fallback: Making<F>,
+    _p: PhantomData<fn(A)>,
+}
+
+pub struct ResponseFuture<P, F, A>
+where
+    P: Future<Error = Error<A>>,
+    F: svc::Service<http::Request<A>>,
+{
+    fallback: F,
+    state: State<P, F::Future>,
+}
+
+pub enum Body<A, B> {
+    A(A),
+    B(B),
+}
+
+enum State<P, F> {
```

Comment:
prefer `ResponseInner` or something that makes it clearer how/where it's used.. `State` would be fine if this was in a file with only the responsefuture...

Reply:
personally i felt like "state" conveys at least slightly more meaning than "inner" --- how about `ResponseState`?